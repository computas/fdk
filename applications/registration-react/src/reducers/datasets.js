import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../constants/ActionTypes';

export default function datasets(
  state = { isFetchingDatasets: false, datasetItems: null },
  action
) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        isFetchingDatasets: true
      };
    case DATASETS_SUCCESS:
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: action.response.data
      };
    case DATASETS_FAILURE:
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: null
      };
    default:
      return state;
  }
}
