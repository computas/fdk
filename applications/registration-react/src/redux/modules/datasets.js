import _ from 'lodash';
import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../../constants/ActionTypes';
import { fetchActions } from '../fetchActions';

function shouldFetch(metaState) {
  const threshold = 60 * 1000; // seconds
  return (
    !metaState ||
    (!metaState.isFetching &&
      (metaState.lastFetch || 0) < Date.now() - threshold)
  );
}

export function fetchDatasetsIfNeeded(catalogId) {
  return (dispatch, getState) =>
    shouldFetch(_.get(getState(), ['datasets', catalogId, 'meta'])) &&
    dispatch(
      fetchActions(`/catalogs/${catalogId}/datasets?size=100`, [
        { type: DATASETS_REQUEST, meta: { catalogId } },
        { type: DATASETS_SUCCESS, meta: { catalogId } },
        { type: DATASETS_FAILURE, meta: { catalogId } }
      ])
    );
}

const initialState = {};

export default function datasets(state = initialState, action) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: true,
            lastFetch: null
          }
        }
      };
    case DATASETS_SUCCESS:
      return {
        ...state,
        [action.meta.catalogId]: {
          items: _.get(action.payload, ['_embedded', 'datasets']),
          meta: {
            isFetching: false,
            lastFetch: Date.now()
          }
        }
      };
    case DATASETS_FAILURE:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: false,
            lastFetch: null
          }
        }
      };
    default:
      return state;
  }
}

export const getDatasetItemsByCatalogId = (datasets, catalogId) =>
  _.get(datasets, [catalogId, 'items']);
