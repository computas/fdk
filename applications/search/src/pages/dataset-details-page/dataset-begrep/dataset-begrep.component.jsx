import React from 'react';
import PropTypes from 'prop-types';

import { BegrepCollapse } from './begrep-collapse/begrep-collapse.component';
import localization from '../../../lib/localization';
import { getTranslateText } from '../../../lib/translateText';

export class DatasetBegrep extends React.Component {
  // eslint-disable-line react/prefer-stateless-function
  constructor(props) {
    super(props);
    this.state = {
      detailed: false
    };
    this.toggle = this.toggle.bind(this);
  }

  toggle() {
    this.setState({ detailed: !this.state.detailed });
  }

  _renderBegrep() {
    const { subject } = this.props;
    const children = items =>
      items.map(item => {
        if (item.prefLabel && item.definition) {
          return (
            <BegrepCollapse
              key={item.uri}
              prefLabel={getTranslateText(item.prefLabel)}
              definition={getTranslateText(item.definition)}
              note={getTranslateText(item.note)}
              source={item.source}
            />
          );
        }
        return null;
      });
    if (subject) {
      return (
        <div>
          <div className="fdk-container-detail fdk-container-detail-header">
            <i className="fa fa-book fdk-fa-left fdk-color-cta" />
            {localization.dataset.subject}
          </div>
          {children(subject)}
        </div>
      );
    }
    return null;
  }

  _renderKeyword() {
    const { keyword } = this.props;
    const children = items =>
      items.map((item, index) => {
        if (index > 0) {
          return (
            <span key={`dataset-begrep-search-${index}`} className="keyword">
              {`, ${getTranslateText(item)}`}
            </span>
          );
        }
        return (
          <span key={`dataset-begrep-search-${index}`} className="keyword">
            {getTranslateText(item)}
          </span>
        );
      });
    if (keyword) {
      return (
        <div className="fdk-container-detail fdk-container-detail-begrep mt-5">
          <div className="fdk-detail-icon">
            <i className="fa fa-search" />
          </div>
          <div className="fdk-detail-text">
            <h5>{localization.dataset.keyword}</h5>
            <p className="fdk-ingress mb-0">{children(keyword)}</p>
          </div>
        </div>
      );
    }
    return null;
  }

  render() {
    return (
      <section>
        {this._renderBegrep()}
        {this._renderKeyword()}
      </section>
    );
  }
}

DatasetBegrep.defaultProps = {
  subject: null,
  keyword: null
};

DatasetBegrep.propTypes = {
  subject: PropTypes.array,
  keyword: PropTypes.array
};
