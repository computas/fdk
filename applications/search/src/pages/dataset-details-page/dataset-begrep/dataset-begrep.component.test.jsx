import React from 'react';
import { shallow } from 'enzyme';
import { DatasetBegrep } from './dataset-begrep.component';
import datasets from '../../../../test/fixtures/datasets';

test('should render DatasetBegrep correctly with no props', () => {
  const minWrapper = shallow(<DatasetBegrep />);
  expect(minWrapper).toMatchSnapshot();
});

test('should render DatasetBegrep correctly', () => {
  const { _source } = datasets[0];
  const defaultProps = {
    subject: _source.subject,
    keyword: _source.keyword
  };
  const wrapper = shallow(<DatasetBegrep {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render DatasetBegrep correctly with no subject definition', () => {
  const { _source } = datasets[1];
  const defaultProps = {
    subject: _source.subject,
    keyword: _source.keyword
  };
  const wrapper = shallow(<DatasetBegrep {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
