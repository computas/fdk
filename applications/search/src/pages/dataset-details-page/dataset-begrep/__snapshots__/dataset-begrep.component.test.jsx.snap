// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DatasetBegrep correctly 1`] = `
<section>
  <div>
    <div
      className="fdk-container-detail fdk-container-detail-header"
    >
      <i
        className="fa fa-book fdk-fa-left fdk-color-cta"
      />
      Subject
    </div>
    <BegrepCollapse
      definition="alt som er registrert med et organisasjonsnummer "
      key="https://data-david.github.io/Begrep/begrep/Enhet"
      note="Alle hovedenheter, underenheter og organisasjonsledd som er identifisert med et organisasjonsnummer."
      prefLabel="enhet"
      source="https://jira.brreg.no/browse/BEGREP-208"
    />
  </div>
  <div
    className="fdk-container-detail fdk-container-detail-begrep mt-5"
  >
    <div
      className="fdk-detail-icon"
    >
      <i
        className="fa fa-search"
      />
    </div>
    <div
      className="fdk-detail-text"
    >
      <h5>
        Keyword
      </h5>
      <p
        className="fdk-ingress mb-0"
      >
        <span
          className="keyword"
          key="dataset-begrep-search-0"
        >
          Bestemmelse
        </span>
        <span
          className="keyword"
          key="dataset-begrep-search-1"
        >
          , jord
        </span>
        <span
          className="keyword"
          key="dataset-begrep-search-2"
        >
          , regulering
        </span>
        <span
          className="keyword"
          key="dataset-begrep-search-3"
        >
          , statlig bestemmelse
        </span>
      </p>
    </div>
  </div>
</section>
`;

exports[`should render DatasetBegrep correctly with no props 1`] = `<section />`;

exports[`should render DatasetBegrep correctly with no subject definition 1`] = `
<section>
  <div>
    <div
      className="fdk-container-detail fdk-container-detail-header"
    >
      <i
        className="fa fa-book fdk-fa-left fdk-color-cta"
      />
      Subject
    </div>
  </div>
  <div
    className="fdk-container-detail fdk-container-detail-begrep mt-5"
  >
    <div
      className="fdk-detail-icon"
    >
      <i
        className="fa fa-search"
      />
    </div>
    <div
      className="fdk-detail-text"
    >
      <h5>
        Keyword
      </h5>
      <p
        className="fdk-ingress mb-0"
      >
        <span
          className="keyword"
          key="dataset-begrep-search-0"
        >
          Bestemmelse
        </span>
        <span
          className="keyword"
          key="dataset-begrep-search-1"
        >
          , jord
        </span>
        <span
          className="keyword"
          key="dataset-begrep-search-2"
        >
          , regulering
        </span>
        <span
          className="keyword"
          key="dataset-begrep-search-3"
        >
          , statlig bestemmelse
        </span>
      </p>
    </div>
  </div>
</section>
`;
